src/coreclr/gc/gc.cpp

```diff
@@ -6885,11 +7032,26 @@ void gc_heap::gc_thread_function ()
 #endif //BACKGROUND_GC
 
 #ifdef MULTIPLE_HEAPS
+            int lowest_heap_with_msl_uoh = -1;
             for (int i = 0; i < gc_heap::n_heaps; i++)
             {
                 gc_heap* hp = gc_heap::g_heaps[i];
-                hp->add_saved_spinlock_info (false, me_release, mt_block_gc);
-                leave_spin_lock(&hp->more_space_lock_soh);
+                if (spin_lock_taken_p (&hp->more_space_lock_soh))
+                {
+                    hp->add_saved_spinlock_info (false, me_release, mt_block_gc, msl_entered);
+                    leave_spin_lock(&hp->more_space_lock_soh);
+                }
+
+                if ((lowest_heap_with_msl_uoh == -1) && (hp->uoh_msl_before_gc_p))
+                {
+                    lowest_heap_with_msl_uoh = i;
+                }
+
+                if (hp->uoh_msl_before_gc_p)
+                {
+                    dprintf (5555, ("h%d uoh msl was taken before GC", i));
+                    hp->uoh_msl_before_gc_p = false;
+                }
```

Comment:
shouldn't this code be under `#ifdef STRESS_DYNAMIC_HEAP_COUNT`?

Reply:
Yes, it should be.